### PR TITLE
[Mono] Fix function pointer check

### LINKED ISSUE #80095: ArrayTypeMismatchException in nested arrays with function pointers (android)
### Description

The code below throws `System.ArrayTypeMismatchException` when run in Android emulator (MAUI project targeting Android).
It works fine in windows.

### Reproduction Steps

```csharp
unsafe class Program
{
    public static delegate* managed<int>[][] Functions = {
        new delegate* managed<int>[]
        {
            &Function,
        },
    };

    public static int Function() => 42;
}
```

### Expected behavior

No exception

### Actual behavior

System.ArrayTypeMismatchException: 'Attempted to access an element as a type incompatible with the array.'

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

Host:
  Version:      7.0.1
  Architecture: x64
  Commit:       97203d38ba

OS Name:     Windows
 OS Version:  10.0.19044
 OS Platform: Windows

### Other information

_No response_

### PR DESCRIPTION
**Pull request by fanyang-mono**
Fix #80095

This PR fixed the function pointer checks for both interpreter and JIT. I also added runtime tests to lock down this behavior.

This fix needs to be backported to .NET7 when it gets approved.

### CODE REVIEW

File: src/mono/mono/metadata/marshal.c [hunk 1]
```diff
@@ -4702,7 +4702,7 @@ get_virtual_stelemref_kind (MonoClass *element_class)
 	if (element_class == mono_defaults.object_class)
 		return STELEMREF_OBJECT;
 	if (is_monomorphic_array (element_class))
-		return STELEMREF_SEALED_CLASS;
+		return STELEMREF_COMPLEX;
```
Comment (lambdageek) [inline src/mono/mono/metadata/marshal.c:4705]: This is going to make all array-of-array checks go to the slow isinst check, isn't it?

Can we make `is_monomorphic_array` more sensitive instead: maybe if its element is a function pointer, then return false?

Comment (fanyang-mono) [reply src/mono/mono/metadata/marshal.c]: Fixed.

**Review by lambdageek — APPROVED** [review]
Thanks!

File: src/tests/baseservices/typeequivalence/functionpointer/Functionpointer.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
+using System;
```
Comment (lambdageek) [inline src/tests/baseservices/typeequivalence/functionpointer/Functionpointer.cs:1]: @fanyang-mono nit: "typeequivalence" not general "type equivalence". it's a special thing that CoreCLR implements for COM interop https://learn.microsoft.com/en-us/dotnet/framework/interop/type-equivalence-and-embedded-interop-types.

A better place for this test would be in src/tests/Loader/classloader/Casting maybe

Comment (fanyang-mono) [reply src/tests/baseservices/typeequivalence/functionpointer/Functionpointer.cs:1]: Ah! Will move it.

**Comment by fanyang-mono** [comment]
Similar crash seen on 'Build Libraries Test Run release mono linux arm64 Debug' showed up on rolling build as well. Not related to this PR.

**Comment by fanyang-mono** [comment]
Verified that newly added test ran on CI. Inspected CI lane `Build osx-x64 Release AllSubsets_Mono_Interpreter_RuntimeTests monointerpreter`. For work item `Loader.classloader`, where new test was added to, in this PR, there were 79 tests being run. (`Loader.classloader.XUnitWrapper  Total: 79, Errors: 0, Failed: 0, Skipped: 0, Time: 8.450s`). For another currently open PR, there were 78 tests being run. (`Loader.classloader.XUnitWrapper  Total: 78, Errors: 0, Failed: 0, Skipped: 0, Time: 8.849s`)

**Pull request merged**